This patch file is from Qualcomm Technologies, Inc.
and is provided under the following license:
Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
SPDX-License-Identifier: BSD-3-Clause-Clear
From a8a14f600760eb7930a700236e2bca31b011dc5a Mon Sep 17 00:00:00 2001
From: Sowmika S <sowms@qti.qualcomm.com>
Date: Mon, 22 Jun 2026 13:09:56 +0000
Subject: [PATCH] Enable SSH gateway IPv4/IPv6 support for echo and fix empty
 WAN interface checks in firewall.

Signed-off-by: Sowmika S <sowms@qti.qualcomm.com>
---
 source/firewall/firewall.c                    |  4 ++--
 source/firewall/firewall_ipv6.c               |  2 +-
 source/scripts/init/service.d/service_sshd.sh | 23 ++++++++++++++-----
 3 files changed, 20 insertions(+), 9 deletions(-)

Index: git/source/firewall/firewall.c
===================================================================
--- git.orig/source/firewall/firewall.c
+++ git/source/firewall/firewall.c
@@ -6906,7 +6906,7 @@ static int do_wan2self_ports(FILE *mangl
         // FIREWALL_DEBUG("Entering do_wan2self_ports\n");    
    // since connection tracking is turned of if current_wan_ipaddr = 0.0.0.0
    // we need to explicitly allow dns
-   if (!isWanReady) {
+   if ((!isWanReady) && ('\0' != default_wan_ifname[0])) {
       fprintf(filter_fp, "-A wan2self_ports -i %s -p udp -m udp --sport 53 -j xlog_accept_wan2self\n", default_wan_ifname);
    }
 
@@ -12470,7 +12470,7 @@ static int prepare_subtables(FILE *raw_f
             {
               fprintf(filter_fp, "-A INPUT -i %s -p tcp -m tcp --dport 22 -j SSH_FILTER\n", current_wan_ifname);
             }
-            else
+            else if('\0' != default_wan_ifname[0])
             {
               fprintf(filter_fp, "-A INPUT -i %s -p tcp -m tcp --dport 22 -j SSH_FILTER\n", default_wan_ifname);
             }
Index: git/source/firewall/firewall_ipv6.c
===================================================================
--- git.orig/source/firewall/firewall_ipv6.c
+++ git/source/firewall/firewall_ipv6.c
@@ -566,7 +566,7 @@ void do_ipv6_filter_table(FILE *fp){
    do_block_ports(fp);	
    fprintf(fp, ":%s - [0:0]\n", "LOG_SSH_DROP");
    fprintf(fp, ":%s - [0:0]\n", "SSH_FILTER");
-   if(bEthWANEnable)
+   if((bEthWANEnable) && ('\0' !=current_wan_ifname[0]))
    {
    fprintf(fp, "-A INPUT -i %s -p tcp -m tcp --dport 22 -j SSH_FILTER\n",current_wan_ifname);
    }
Index: git/source/scripts/init/service.d/service_sshd.sh
===================================================================
--- git.orig/source/scripts/init/service.d/service_sshd.sh
+++ git/source/scripts/init/service.d/service_sshd.sh
@@ -75,7 +75,10 @@ fi
 SERVICE_NAME="sshd"
 SELF_NAME="`basename "$0"`"
 
-PID_FILE=/var/run/dropbear.pid
+#PID_FILE=/var/run/dropbear.pid
+if [ "$BOX_TYPE" = "echo" ]; then
+      . /etc/utopia/service.d/service_sshd_qti.sh
+fi
 PMON=/etc/utopia/service.d/pmon.sh
 if [ -f /etc/mount-utils/getConfigFile.sh ];then
       mkdir -p /tmp/.dropbear
@@ -270,6 +273,8 @@ do_start() {
        if  ([ "$MANUFACTURE" = "Technicolor" ] || [ "$MODEL_NUM" = "SG417DBCT" ] || [ "$BOX_TYPE" = "WNXL11BWL" ]) ; then
 	      echo_t "dropbear -E -s -K 60 -b /etc/sshbanner.txt ${commandString} -r ${DROPBEAR_PARAMS_1} -r ${DROPBEAR_PARAMS_2} -a -P ${PID_FILE}"
           dropbear -E -s -b /etc/sshbanner.txt $commandString -r $DROPBEAR_PARAMS_1 -r $DROPBEAR_PARAMS_2 -a -P $PID_FILE -K 60 $USE_DEVKEYS 2>>$CONSOLEFILE
+       elif ([ "$BOX_TYPE" = "echo" ]); then
+              start_dropbear_echo
        elif [ "$BOX_TYPE" = "SCER11BEL" -a "$LANIPV6Support" = "true" ]; then
               dropbear -E -s -b /etc/sshbanner.txt $commandString -r $DROPBEAR_PARAMS_1 -r $DROPBEAR_PARAMS_2 -a -P $PID_FILE -K 60 $USE_DEVKEYS 2>>$CONSOLEFILE
        else
@@ -296,8 +301,9 @@ do_stop() {
 #   kill -9 dropbear
    if [ ! -f "$PID_FILE" ] ; then
      tmp_filename="/tmp/dropbear.pid"
-     #Get the dropbear IPV6 process IDs. There could be more than 1 dropbear IPV6 process running based on number of open ssh connections
-     ps | grep 'dropbear -E -s -b /etc/sshbanner.txt' | sed '/grep/d' > $tmp_filename
+     if [ "$BOX_TYPE" = "echo" ]; then
+       stop_dropbear_echo $tmp_filename
+     fi
      while read -r line; do
        pid=`echo "$line" | head -n1 |  awk '{print $1;}'`
        kill -9 "$pid"
@@ -356,7 +362,8 @@ service_start() {
    #dropbear process would be running on demand basis
    if [ "$BOX_TYPE" != "HUB4" ] && [ "$BOX_TYPE" != "SR300" ] && [ "$BOX_TYPE" != "SE501" ] && [ "$BOX_TYPE" != "WNXL11BWL" ] && [ "$BOX_TYPE" != "SR213" ] &&  [ "$BOX_TYPE" != "SCER11BEL" ] && [ "$BOX_TYPE" != "SCXF11BFL" ]; then
 
-                $PMON setproc ssh dropbear $PID_FILE "/etc/utopia/service.d/service_sshd.sh sshd-restart"
+                $PMON setproc ssh dropbear $PID_FILE1 "/etc/utopia/service.d/service_sshd.sh sshd-restart"
+                $PMON setproc ssh dropbear $PID_FILE2 "/etc/utopia/service.d/service_sshd.sh sshd-restart"
     fi
 		sysevent set ${SERVICE_NAME}-errinfo
 		sysevent set ${SERVICE_NAME}-status "started"
@@ -449,8 +456,12 @@ case "$1" in
       service_bridge_status
       ;;
   current_wan_ifname)
-      service_stop
-      service_start
+      SSHACCESS=`syscfg get mgmt_wan_sshaccess`
+      LAN_SSHACCESS=`syscfg get mgmt_lan_sshaccess`
+      if [ "$SSHACCESS" == "1" ] ||  [ "$LAN_SSHACCESS" == "1" ]; then
+         service_stop
+         service_start
+      fi
       ;;
 
   *)
